### PR TITLE
Move listener stream construction out of `serve`

### PR DESCRIPTION
**Pull request by andrisaar**
Tentative step towards possibly having multiple kinds of streams, e.g. virtio-vsock streams.

The amount of generic arguments `serve` has is starting to get rather silly, though.

### CODE REVIEW

File: oak_functions_containers_app/src/lib.rs [hunk 1]
```diff
@@ -390,15 +391,18 @@ const GRPC_STATUS_HEADER_CODE: &str = "grpc-status";
 const MAX_DECODING_MESSAGE_SIZE: usize = 1024 * 1024 * 1024;
 
 // Starts up and serves an OakFunctionsContainersService instance from the provided TCP listener.
-pub async fn serve<G, H: Handler>(
-    listener: TcpListener,
+pub async fn serve<G, H, S, IO, IE>(
```
Comment (tiziano88) [inline oak_functions_containers_app/src/lib.rs:394]: I think most of these generics don't actually need to be generics. In general, if `Box<dyn T>` works instead, I think we should go with that, unless we absolutely need to keep things on the stack. WDYT?

Comment (andrisaar) [reply oak_functions_containers_app/src/lib.rs:394]: Thus far the general guidance I've got has been the opposite, prefer not to `Box` things if possible :)

In this case, we call `Router::serve_with_incoming()`, and I think that one wants a `Stream` and won't work with a boxed value.

Comment (tiziano88) [reply oak_functions_containers_app/src/lib.rs:394]: Let's continue the discussion on http://b/328114871 .

Comment (andrisaar) [reply oak_functions_containers_app/src/lib.rs:394]: Yep, we can continue bikeshedding there, but I'd argue this PR is still worth submitting :)

**Review by tiziano88 — APPROVED** [review]

**Pull request merged**